compiler/codegen/symbol.cpp

```diff
@@ -2165,6 +2172,114 @@ void FnSymbol::codegenPrototype() {
   return;
 }
 
+namespace {
+  /* This little analysis helps with llvm.loop.parallel_accesses
+     metadata.
+
+     It identifies references that are defined inside an order-independent
+     loop. If it can show that these references refer to something outside
+     of any order-independent loop in the function, it marks them with
+     FLAG_POINTS_NON_STACK.
+
+     The goal here is to avoid hinting llvm.loop.parallel_accesses
+     for load/store to temporary variables that are declared inside
+     of an order independent loop. The reason for that is that the LLVM
+     optimizers do not yet handle vectorizing allocas. By not marking
+     load/store with allocas for variables within the loop as
+     parallel_accesses, vectorization will be inhibited unless these
+     are all promoted to registers (or the LLVM vectorization pass can
+     prove it is OK for some other reason).
+   */
+  struct MarkNonStackVisitor : public AstVisitorTraverse {
+    LoopStmt* outerMostOrderIndependentLoop;
+    MarkNonStackVisitor() : outerMostOrderIndependentLoop(NULL) { }
+    void handleLoopStmt(LoopStmt* loop);
+    bool exprPointsToNonStack(Expr* e);
+    bool enterCallExpr(CallExpr* call);
+    bool enterWhileDoStmt(WhileDoStmt* loop);
+    bool enterDoWhileStmt(DoWhileStmt* loop);
+    bool enterCForLoop(CForLoop* loop);
+    bool enterForLoop(ForLoop* loop);
+  };
+}
+
+void MarkNonStackVisitor::handleLoopStmt(LoopStmt* loop) {
+  if (loop->isOrderIndependent() && outerMostOrderIndependentLoop == NULL) {
+    outerMostOrderIndependentLoop = loop;
+  }
+}
+
+bool MarkNonStackVisitor::exprPointsToNonStack(Expr* e) {
+  if (SymExpr* se = toSymExpr(e)) {
+    Symbol* sym = se->symbol();
+
+    if (sym->hasFlag(FLAG_POINTS_NON_STACK))
+      return true; // already marked
+
+    if (outerMostOrderIndependentLoop != NULL &&
+        !outerMostOrderIndependentLoop->contains(sym->defPoint)) {
+      // defined outside the outermost independent loop
+      // (could be a global variable or argument, too)
+      return true;
+    }
+  }
+
+  if (CallExpr* call = toCallExpr(e)) {
+    FnSymbol* fn = call->resolvedOrVirtualFunction();
+    if (call->isPrimitive(PRIM_GET_MEMBER) ||
+        call->isPrimitive(PRIM_GET_SVEC_MEMBER) ||
```

Comment:
I don't think we care about the `_VALUE` variants here because the analysis is marking `ref`s. Consider getting `foo.bar` with a `_VALUE` version:
 * if `bar` is itself a ref, then we don't know what it refers to, so can't mark it
 * if `bar` is not a ref, then we will read its value (say, an `int`), and don't need to mark it

Reply:
But inside this `if`, we mark `call->get(1)` (foo) and not `call->get(2)` (bar)?

Reply:
The inside of the if doesn't mark anything? It is checking if `foo` is a class; if so, the reference to `foo.bar` points to heap memory.

Reply:
Maybe I am confused because of a misunderstanding of what `_VALUE` versions do.

To put my question differently, if `foo` is a class, don't we read something from heap (foo) when doing `foo.bar` even if we are reading `bar` by value?

Reply:
Yes. But that case is handled in expr.cpp rather than in the analysis here.

Reply:
OK, then. Thanks!